src/common/gamut_mapping.c

```diff
@@ -0,0 +1,333 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2022 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#include "chromatic_adaptation.h"
+#include "gamut_mapping.h"
+#include "iop/gaussian_elimination.h"
+
+
+#define TOLERANCE 1e-6f
+
+
+static float find_max_L(const float a, const float b, const float max_L, const float white_luminance, const dt_colormatrix_t output_matrix)
+{
+  // Find the upper boundary maximum lightness at given chroma by bisection
+  float upper = max_L;
+  float lower = 0.f;
+  float midpoint = (upper + lower) / 2.f;
+  dt_aligned_pixel_t RGB;
+  do
+  {
+    dt_aligned_pixel_t XYZ_D65;
+    dt_aligned_pixel_t Lab = { midpoint, a, b, 0.f };
+    Oklab_to_XYZ_D65(Lab, XYZ_D65);
+    dt_apply_transposed_color_matrix(XYZ_D65, output_matrix, RGB);
+    if(RGB[0] < white_luminance && RGB[1] < white_luminance && RGB[2] < white_luminance) lower = midpoint;
+    else upper = midpoint;
+    midpoint = (upper + lower) / 2.f;
+  } while(upper - lower > TOLERANCE);
+  return midpoint;
+}
+
+
+static void sample_upper_boundary(const float a, const float b,
+                                  const float white_lightness, const float white_luminance,                                        
+                                  const dt_colormatrix_t output_matrix,
+                                  float samples[][2])
+{
+  // Take a number of samples from the upper boundary
+  const float C_step = 1.f / GAMUT_MAP_UPPER_SAMPLES;
+  float max_L = 1.f;
+  for(size_t i = 0; i < GAMUT_MAP_UPPER_SAMPLES; i++)
+  {
+    const float C = (i + 1) * C_step;
+    max_L = find_max_L(C * a, C * b, max_L, white_luminance, output_matrix);
+    samples[i][0] = max_L;
+    samples[i][1] = C;
+  }
+}
+
+
+static float find_max_C(const float a, const float b, const float L, const float min_C, const float black_luminance, const dt_colormatrix_t output_matrix)
+{
+  // Find the lower boundary maximum chroma at given lightness by bisection
```

Comment:
Given that the gamut is a double cone, its boundary can be modelled by saturation = chroma / lightness for the bottom part. 
Then, for the upper part, you find the line of (chroma / (white - lightness)).

Reply:
In general the gamut is not necessarily a simple double cone constrained by straight lines at each slice (for e.g. JzAzBz, both upper and lower boundary are curved). Therefore I intentionally made the gamut boundary handling a little more generic than it needs to be for Oklab, to allow for experimentation with other color spaces.

For Oklab the lower boundary seems to be a straight line but the upper one is slightly curved, therefore the 5th degree polynomial.
![gamut_slice](https://user-images.githubusercontent.com/5001906/150766635-b64d4936-04d2-4be4-b886-86d5c7845a11.png)

See also @bottosson's excellent [post](https://bottosson.github.io/posts/gamutclipping/#gamut-clipping) where there are more images of hue slices in Oklab.

Reply:
Ok thanks